### PR TITLE
Restrict access to user to specific teams group

### PR DESCRIPTION
**Pull request by sintetico82**
I try to solve this issue "Need restriction on account domain #5". 
I read the user teams information from Bitbucket API and check if they are in the list of teams in configuration of Sonarqube. I also upgrade dependance in the POM for sonar-plugin-api to version 5.5 becouse I needed to use Sonarqube class   `org.sonar.api.server.authentication.UnauthorizedException`.

I hope it can be usefull.

### CODE REVIEW

**Comment by starchx** [comment]
Waiting for this feature as well. Just to double check, if I have 'abcde' as my bitbucket team ID, does that mean I need to create a Group in SonarQube with exact name as 'abcde' to match? 
or better if we are allow to configure list of teams to allow somewhere in SonarQube?

**Comment by sintetico82** [comment]
Yes of course, You can configure a list of teams in configuration section of SonqrQube. I attached a screenshot of how look like the configuration. You can specific more groups with comma separation like "abcde,fghi"

sonarqubebitbucketteamaccess

**Comment by sintetico82** [comment]
Sorry, nobody can check it for merge it?

**Comment by shuraa** [comment]
@simonbrandhof , hi! can you check and merge this PR :) ?

File: src/main/java/org/sonarqube/auth/bitbucket/BitbucketIdentityProvider.java [hunk 1]
```diff
@@ -133,6 +156,18 @@ private GsonEmails requestEmails(OAuthService scribe, Token accessToken) {
...
 
+  @CheckForNull
+  private GsonTeams requestTeams(OAuthService scribe, Token accessToken) {
+    OAuthRequest userRequest = new OAuthRequest(Verb.GET, settings.apiURL() + "2.0/teams?role=member", scribe);
+    scribe.signRequest(accessToken, userRequest);
+    Response teamsResponse = userRequest.send();
+    if (teamsResponse.isSuccessful()) {
+      return GsonTeams.parse(teamsResponse.getBody());
+    } else 
+       LOGGER.warn( format("Error to retrive Bitbucket teams %s", teamsResponse.getBody()));
```
Comment (julienlancelot) [inline src/main/java/org/sonarqube/auth/bitbucket/BitbucketIdentityProvider.java]: Please add {} around this line

File: src/main/java/org/sonarqube/auth/bitbucket/BitbucketIdentityProvider.java [hunk 2]
```diff
@@ -103,6 +110,22 @@ public void callback(CallbackContext context) {
 
     GsonUser gsonUser = requestUser(scribe, accessToken);
     GsonEmails gsonEmails = requestEmails(scribe, accessToken);
+
+    String[] teamsRestriction = settings.teamRestriction();
```
Comment (julienlancelot) [inline src/main/java/org/sonarqube/auth/bitbucket/BitbucketIdentityProvider.java]: Please extract all this code in a specific method (something like `checkTeamRestriction(scribe, accessToken)`), to increase readability.

File: src/main/java/org/sonarqube/auth/bitbucket/BitbucketIdentityProvider.java [hunk 3]
```diff
@@ -103,6 +110,22 @@ public void callback(CallbackContext context) {
...
+    if (teamsRestriction != null && teamsRestriction.length > 0) {
+
+      GsonTeams gsonTeams = requestTeams(scribe, accessToken);
+      if (gsonTeams == null || gsonTeams.getTeams() == null || gsonTeams.getTeams().size() == 0)
+        throw new UnauthorizedException(format("No teams found for current user: %s", gsonUser.getUsername()));
+
+      if (!gsonTeams.getTeams().stream().anyMatch(t -> Arrays.asList(teamsRestriction).contains(t.getUserName()))) {
+          LOGGER.trace(format("User %s is not part of restricted teams: '%s'", gsonUser.getUsername(),
```
Comment (julienlancelot) [inline src/main/java/org/sonarqube/auth/bitbucket/BitbucketIdentityProvider.java]: I'm not sure this log is really bringing value, as the message is returned to the user.

**Comment by julienlancelot** [comment]
Please also add some tests that cover the new feature in `IntegrationTest`

**Comment by sintetico82** [comment]
Sorry, i try to merge the master branch to my feature branch, but i have this error from maven:

`[ERROR] Failed to execute goal on project sonar-auth-bitbucket-plugin: Could not resolve dependencies for project org.sonarsource.auth.bitbucket:sonar-auth-bitbucket-plugin:sonar-plugin:1.1-SNAPSHOT: Failure to find org.sonarsource.sonarqube:sonar-plugin-api:jar:7.2.0.12315 in https://repo.maven.apache.org/maven2 was cached in the local repository, resolution will not be reattempted until the update interval of central has elapsed or updates are forced -> [Help 1]`

**Comment by julienlancelot** [comment]
@sintetico82 The SonarQube version has been fixed, it's now 7.2

**Comment by sintetico82** [comment]
The endless integration...

**Comment by simonbrandhof** [comment]
Most of feedbacks have being applied in the branch [sintetico82-feature/restrict-access-teams](https://github.com/SonarSource/sonar-auth-bitbucket/tree/sintetico82-feature/restrict-access-teams). Work is in progress.

**Comment by simonbrandhof** [comment]
Closing this pull request in favour of #22. Thanks for you contributions.

**Pull request closed without merging**